File: day25/src/main.py
from pathlib import Path

def part1():
    data = Path('my_input.txt').read_text().strip().split("\n\n")
    locks, keys = [], []

    for schematic in data:
        rows = schematic.splitlines()
        heights = [sum(row[col] == "#" for row in rows) - 1 for col in range(len(rows[0]))]
        (locks if all(cell == "#" for cell in rows[0]) else keys).append(heights)

    fitting_pairs_count = sum(
        all(l + k <= 5 for l, k in zip(lock, key))
            for lock in locks
            for key in keys
    )

    print(f'Answer: {fitting_pairs_count}')

if __name__ == '__main__':
    part1()

File: day25/src/main.rs
use std::fs;

fn part1() {
    let data = fs::read_to_string("my_input.txt").expect("Failed to read file");
    let mut locks = Vec::new();
    let mut keys = Vec::new();

    for schematic in data.trim().split("\n\n") {
        let rows: Vec<&str> = schematic.lines().collect();
        let heights: Vec<usize> = (0..rows[0].len())
            .map(|col| rows.iter().filter(|row| row.chars().nth(col) == Some('#')).count() - 1)
            .collect();

        if rows[0].chars().all(|c| c == '#') { locks.push(heights) } else { keys.push(heights) }
    }

    let fitting_pairs_count: usize = locks
        .iter()
        .map(|lock| {
            keys.iter()
                .filter(|key| lock.iter().zip(key.iter()).all(|(l, k)| l + k <= 5))
                .count()
        })
        .sum();

    println!("Answer: {}", fitting_pairs_count);
}

fn main() {
    part1();
}
